Core/MaterialExtension.cpp

```diff
@@ -16,66 +16,103 @@
   fTarget = false ;
   //! depends on the material typ
-  fPickOffFraction = 0.;
+  Material = new MaterialParameters();
+  Material->SetMaterialByName(materialType);
+  FillIntensities();
 }
 
-
 MaterialExtension::~MaterialExtension() {}
 
-const G4double MaterialExtension::Get3gFraction()
+void MaterialExtension::Add_oPsComponent( G4double lifetime, G4double probability )
 {
-  return foPsFraction;
+  if(lifetime < 0. * ns || lifetime > MaterialParameters::oPsTauVaccum) 
+    G4Exception("MaterialExtension", "ME01", JustWarning, "given oPs lifetime not in the range [0, 142] ns"); 
+  else if( probability < 0. || probability > 100.)
+    G4Exception("MaterialExtension", "ME01", JustWarning, "given oPs probability not in the range [0, 100] %"); 
+  else 
+    Material->Add_oPsComponent(lifetime, probability);
 }
 
-void MaterialExtension::SetoPsFraction(G4double x)
+void MaterialExtension::Add_directComponent( G4double lifetime, G4double probability )
 {
-  if (x < 0 || x > 1) {
-    G4Exception(
-      "MaterialExtension", "ME01", JustWarning, "3gamma fraction can not extend 1; nothing happend"
-    );
-  } else {
-    foPsFraction = x;
-  };
+  if(lifetime < 0. * ns || lifetime > MaterialParameters::directTauMax) 
+    G4Exception("MaterialExtension", "ME01", JustWarning, "given direct lifetime not in the range [0, 0.6] ns"); 
+  else if( probability < 0. || probability > 100.)
+    G4Exception("MaterialExtension", "ME01", JustWarning, "given direct probability not in the range [0, 100] %"); 
+  else 
+    Material->Add_DirectComponent(lifetime, probability);
 }
 
-void MaterialExtension::SetoPsLifetime(G4double x)
+void MaterialExtension::Set_pPsComponent(G4double lifetime, G4double fraction)
 {
-  if (x < 0 || x > MaterialParameters::oPsTauVaccum) {
-    G4Exception(
-      "MaterialExtension", "ME02", JustWarning, "oPs lifetime can not excide 142*ns"
-    );
-  } else {
-    foPslifetime = x;
-  };
+  if( lifetime < 0. * ns || lifetime > 0.2 ) 
+    G4Exception("MaterialExtension", "ME01", JustWarning, "given pPs lifetime not in the range [0, 0.2] ns"); 
+  else if( fraction < 0. || fraction > 1.)
+    G4Exception("MaterialExtension", "ME01", JustWarning, "given pPs fraction not in the range [0, 1]"); 
+  else 
+    Material->Set_pPsComponent(lifetime, fraction);
 }
 
-const std::vector<G4double> MaterialExtension::GetEventsFraction()
+const G4double MaterialExtension::GetLifetime( double randNumber, G4String channel )
 {
-  G4double direct3g = MaterialParameters::fractionDirect3g;
+  if( channel == "ortho2G" )
+    return Material->GetoPs2GLifetimeFromVector(randNumber);
+  else if( channel == "ortho3G" )
+    return Material->GetoPs3GLifetimeFromVector(randNumber);  
+  else if( channel == "para2G" )
+    return Material->GetpPsLifetime();
+  else
+    return Material->GetDirectLifetimeFromVector(randNumber);
+}
 
-  if (foPsFraction == 1) {
-    direct3g = 0.0f;
-  }
+void MaterialExtension::ChangeMaterialConstants()
```

Comment:
that is a bit unexpected. Maybe some comment what it does?

Reply:
I made a dummy object MaterialParameters::Temp, 
for which every lifetime and intensity parameter can be modified by the user in the MaterialExtensionMessenger class (from macro file).
This function is changing Material (xad4, aluminium, kapton...) lifetime parameters to the lifetime parameters that are currently in the MaterialParameters::Temp

Also, after ChangeMaterialConstants(), MaterialParameters::Temp is cleared, in the MaterialExtensionMessenger.

Reply:
Ok.  Thank you. Maybe you could add a similar explanation somewhere in some comments.